client.go

```diff
@@ -52,99 +59,70 @@ func SelectIfaces(ifaces []net.Interface) ClientOption {
 	}
 }
 
-// Resolver acts as entry point for service lookups and to browse the DNS-SD.
-type Resolver struct {
-	c *client
-}
-
-// NewResolver creates a new resolver and joins the UDP multicast groups to
-// listen for mDNS messages.
-func NewResolver(options ...ClientOption) (*Resolver, error) {
-	// Apply default configuration and load supplied options.
-	var conf = clientOpts{
-		listenOn: IPv4AndIPv6,
-	}
-	for _, o := range options {
-		if o != nil {
-			o(&conf)
-		}
-	}
-
-	c, err := newClient(conf)
+// Browse for all services of a given type in a given domain.
+func Browse(ctx context.Context, service, domain string, entries chan<- *ServiceEntry, opts ...ClientOption) error {
+	cl, err := newClient(applyOpts(opts...))
 	if err != nil {
-		return nil, err
+		return err
 	}
-	return &Resolver{
-		c: c,
-	}, nil
-}
-
-// Browse for all services of a given type in a given domain.
-func (r *Resolver) Browse(ctx context.Context, service, domain string, entries chan<- *ServiceEntry) error {
 	params := defaultParams(service)
 	if domain != "" {
 		params.Domain = domain
 	}
 	params.Entries = entries
 	params.isBrowsing = true
-	ctx, cancel := context.WithCancel(ctx)
-	go r.c.mainloop(ctx, params)
+	return cl.run(ctx, params)
+}
 
-	err := r.c.query(params)
+// Lookup a specific service by its name and type in a given domain.
+func Lookup(ctx context.Context, instance, service, domain string, entries chan<- *ServiceEntry, opts ...ClientOption) error {
+	cl, err := newClient(applyOpts(opts...))
 	if err != nil {
-		cancel()
 		return err
 	}
-	// If previous probe was ok, it should be fine now. In case of an error later on,
-	// the entries' queue is closed.
-	go func() {
-		if err := r.c.periodicQuery(ctx, params); err != nil {
-			cancel()
-		}
-	}()
-
-	return nil
-}
-
-// Lookup a specific service by its name and type in a given domain.
-func (r *Resolver) Lookup(ctx context.Context, instance, service, domain string, entries chan<- *ServiceEntry) error {
 	params := defaultParams(service)
 	params.Instance = instance
 	if domain != "" {
 		params.Domain = domain
 	}
 	params.Entries = entries
-	ctx, cancel := context.WithCancel(ctx)
-	go r.c.mainloop(ctx, params)
-	err := r.c.query(params)
-	if err != nil {
-		// cancel mainloop
-		cancel()
-		return err
+	return cl.run(ctx, params)
```

Comment:
This blocks now, right? Did it used to block?

Reply:
It does now, but it didn't before.

Reply:
I see. We should definitely document that as users will need to read from the channel in a separate goroutine.

Reply:
Done.